roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -0,0 +1,247 @@
+  * Keep track of an `all_approved_max: Option<Hash>`.
+  * For each block hash encountered, load the `BlockEntry` associated. If any are not found, return `None` on the response channel and conclude.
+  * If the block entry's `approval_bitfield` has all bits set to 1 and `all_approved_max == None`, set `all_approved_max = Some(current_hash)`.
+  * If the block entry's `approval_bitfield` has any 0 bits, set `all_approved_max = None`.
+  * After iterating all ancestry, return `all_approved_max`.
+
+### Utility
+
+#### `import_checked_assignment`
+  * Load the candidate in question and access the `approval_entry` for the block hash the cert references.
+  * Ensure the validator index is not part of the backing group for the candidate.
+  * Ensure the validator index is not present in the approval entry already.
+  * Create a tranche entry for the delay tranche in the approval entry and note the assignment within it.
+  * Note the candidate index within the approval entry.
+
+#### `import_checked_approval(BlockEntry, CandidateEntry, ValidatorIndex)`
+  * Set the corresponding bit of the `approvals` bitfield in the `CandidateEntry` to `1`.
+  * For each `ApprovalEntry` in the `CandidateEntry` (typically only 1), check whether the validator is assigned as a checker.
+    * If so, set `n_tranches = tranches_to_approve(approval_entry)`.
+    * If `check_approval(block_entry, approval_entry, n_tranches)` is true, set the corresponding bit in the `block_entry.approved_bitfield`.
+
+#### `tranches_to_approve(approval_entry) -> tranches`
+  * Determine the amount of tranches `n_tranches` our view of the protocol requires of this approval entry
+    * First, take tranches until we have at least `session_info.needed_approvals`. Call the number of tranches taken `k`
+    * Then, count no-shows in tranches `0..k`. For each no-show, we require another checker. Take new tranches until each no-show is covered, so now we've taken `l` tranches. e.g. if there are 2 no-shows, we might only need to take 1 additional tranche with >= 2 assignments. Or we might need to take 3 tranches, where one is empty and the other two have 1 assignment each.
+    * Count no-shows in tranches `k..l` and for each of those, take tranches until all no-shows are covered. Repeat so on until either
+      * We run out of tranches to take, having not received any assignments past a certain point. In this case we set `n_tranches` to a special value `ALL` which indicates that new assignments are needed.
+      * All no-shows are covered. Set `n_tranches` to the number of tranches taken
+    * return `n_tranches`
+
+#### `check_approval(block_entry, approval_entry, n_tranches) -> bool`
+  * If `n_tranches` is ALL, return false
+  * Otherwise, if all validators in `n_tranches` have approved, return `true`. If any validator in these tranches has not yet approved but is not yet considered a no-show, return `false`.
```

Comment:
I needed `tranches_to_approve` and `check_approval` to be one method because `check_approval` knows about no-shows.